Choose queue type and exchange type when creating missing queues (fix #9671)

ghirailghiro opened the pull request:
## Description

This PR adds **two new optional settings** that control how Celery auto-creates
queues when `task_create_missing_queues = True`:

| Setting | Default | What it does |
|---------|---------|--------------|
| `task_create_missing_queue_type` | `"classic"` | Select the AMQP queue kind: keep the default `"classic"`  or choose the type you need, e.g. `"quorum"`). |
| `task_create_missing_queue_exchange_type` | `None` | Override the exchange type used for the auto-exchange: set to `"topic"`, `"fanout"`, etc. |


## Changes(Fixes #9671)

Add two new settings that apply when Celery autogenerates a queue
(`task_create_missing_queues=True`):

- `task_create_missing_queue_type`

- `task_create_missing_queue_exchange_type`

Backwards compatibility: default behaviour (classic queue + direct exchange) is unchanged.

auvipy reviewed (commented):
lets also fix the kombu front as well as described here (https://github.com/celery/celery/issues/9759#issuecomment-3081079711)

ghirailghiro commented:
Hi @auvipy I just open the PR on Kombu

I added to the control.py file on the celery part the two parameter control_queue_durable and control_queue_exclusive, after kombu is fixed i will make the commit with these modifications.

auvipy commented:
yeah thanks. saw that.

auvipy commented:
can you please also look at the approach in this PR? https://github.com/celery/kombu/pull/2313

ghirailghiro commented:
Hi @auvipy

I’ve added support for configuring durable and exclusive options on both event queues and control queue:

- event_queue_durable, event_queue_exclusive
- control_queue_durable, control_queue_exclusive

There’s also validation to prevent both being set to True, plus I added docs and tests. This is meant to fix #9759 on the celery part.

That said, I can’t fully test the changes because they depend on feature that we introduced in [Kombu PR 2338](https://github.com/celery/kombu/pull/2338)

Just wondering what’s the best way to proceed here, i need some guidance on how i can test it or if is necessary to test with the right version of kombu.

Thanks!

auvipy commented:
can you check why the unit tests are failing, please?

ghirailghiro commented:
My bad, I’ve just pushed a follow‑up commit that adds the missing guard clause to the failing test I introduced.

ghirailghiro commented:
For what i can see now the tests are failing because the tests are not pointing to the main branch of kombu but it's last release. The solution that i propose need to use the new parameters that we merged in the [Kombu PR 2338](https://github.com/celery/kombu/pull/2338) on the Mailbox class.

ghirailghiro commented:
I think the tests are now failing because they are not pointing to the main of Kombu, I think i need some guidance here to understand how to fix them.

auvipy commented:
kombu has been updated, so lets see

auvipy commented:
the redis failures are not related to this rabbitmq bug fix pr

auvipy commented on celery/events/receiver.py:
```diff
@@ -34,25 +35,39 @@ class EventReceiver(ConsumerMixin):
             exchange=self.exchange,
             routing_key=self.routing_key,
-            auto_delete=True, durable=False,
+            auto_delete=True and (not queue_durable),
```
can you cross check this suggestion, please?

ghirailghiro replied:
I think Copilot is right, that True is useless

auvipy commented on celery/events/receiver.py:
```diff
@@ -34,25 +35,39 @@ class EventReceiver(ConsumerMixin):
         self.routing_key = routing_key
         self.node_id = node_id or uuid()
-        self.queue_prefix = queue_prefix or self.app.conf.event_queue_prefix
+        cfg = self.app.conf
```
@Nusnus whats your take on this?

Nusnus replied:
Tbh I think the full `self.app.conf` should be used instead to make the code cleaner.

ghirailghiro replied:
For me that's ok, i will push the modifications

auvipy commented:
@ghirailghiro please address the review comments...

Nusnus reviewed (changes requested):
Nice work!
Very clean PR.

One change and we can merge it, please fix the missing `.. versionadded::` in the docs!

Nusnus commented on docs/userguide/monitoring.rst:
```diff
@@ -814,3 +814,24 @@ worker-offline
+Celery uses `kombu.pidbox.Mailbox` internally to send control and broadcast commands
+to workers.
+
+.. versionadded:: Kombu 5.6.0
```
On second thought, even though we don’t use that notion anywhere else, it does make sense and maybe we should actually adopt mentioning which kombu version introduced a relevant change.

Interesting.

auvipy replied:
yes, this feature is kombu 5.6 dependent---

auvipy commented on docs/userguide/configuration.rst:
```diff
@@ -2619,6 +2623,48 @@ If enabled (default), any queues specified that aren't defined in
 
+.. setting:: task_create_missing_queue_type
+
+``task_create_missing_queue_type``
```
please fix the missing .. versionadded:: annotation in the docs.

ghirailghiro replied:
added to all the new parameters with the last commit

auvipy commented on docs/userguide/configuration.rst:
```diff
@@ -2619,6 +2623,48 @@ If enabled (default), any queues specified that aren't defined in
+.. setting:: task_create_missing_queue_exchange_type
+
+``task_create_missing_queue_exchange_type``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

auvipy commented on celery/app/amqp.py:
```diff
@@ -181,7 +193,21 @@ def deselect(self, exclude):
+        else:
+            exchange = self.autoexchange(name)
+
+        return Queue(name, exchange, name, queue_arguments=queue_arguments)
```
why returning name twice in Queue class?

auvipy commented:
unit tests seems to be failing after the last commit. can you please cross check and fix them?

ghirailghiro commented:
I checked the celery/app/amqp.py on [main](https://github.com/celery/celery/blob/3703beb4589b4d590b84a1b08a8d6558149b6f4d/celery/app/amqp.py#L184C9-L184C15) and i think the right way is to explicit set the routing_key. I think we should roll back the last commit for keeping compatibility.

Nusnus commented:
🚀🚀

ghirailghiro commented:
Thank you guys for the support and sorry for the delay on the response!
